### ISSUE TITLE
Building a Python 3.7 Distroless Image

### ISSUE THREAD

**Issue by HVP1621** [GoogleContainerTools/distroless#371]
For my app, I need Python 3.7 dependencies. I am using Bazel to build a distroless image which runs on top of the existing distroless/base image.
My package sources are based on the Buster snapshots.
The following are my WORKSPACE and BUILD FIles.

### WORKSPACE
```

load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

#Download the rules_docker repository at release v0.7.0
http_archive(
    name = "io_bazel_rules_docker",
    sha256 = "aed1c249d4ec8f703edddf35cbe9dfaca0b5f5ea6e4cd9e83e99f3b0d1136c3d",
    strip_prefix = "rules_docker-0.7.0",
    urls = ["https://github.com/bazelbuild/rules_docker/archive/v0.7.0.tar.gz"],
)

#This is NOT needed when going through the language lang_image
#"repositories" function(s).
load(
    "@io_bazel_rules_docker//repositories:repositories.bzl",
    container_repositories = "repositories",
)
container_repositories()

load(
    "@io_bazel_rules_docker//container:container.bzl",
    "container_pull",
)

container_pull(
  name = "java_base",
  registry = "gcr.io",
  repository = "distroless/java",
)

container_pull(
  name = "gcrbase",
  registry = "gcr.io",
  repository = "distroless/base",
  tag = "debug",
)

####################################
#STARTING TO LOAD PACKAGES FOR THE PYTHON 3.7 IMAGE

load("//package_manager:package_manager.bzl", "package_manager_repositories")
load("//package_manager:dpkg.bzl", "dpkg_list", "dpkg_src")

package_manager_repositories( )

dpkg_src(
    name = "debian_buster",
    arch = "amd64",
    distro = "buster",
    sha256 = "b0a3a1754ce592ec31463f2e893cd11e4f6914ea9d2a6f4bf53ab2e199d3ad2e",
    snapshot = "20190605T025513Z",
    url = "https://snapshot.debian.org/archive",
)

dpkg_src(
    name = "debian_stretch_security",
    package_prefix = "https://snapshot.debian.org/archive/debian-security/20190510T162559Z/",
    packages_gz_url = "https://snapshot.debian.org/archive/debian-security/20190510T162559Z/dists/stretch/updates/main/binary-amd64/Packages.gz",
    sha256 = "0645e7edacb45fa8fec510aa8105bfca2bce155b2a08a4fe9628093dc7405326",
)

dpkg_list(
    name = "package_bundle",
    packages = [
        "libc6=2.28-10",
        "base-files",
        "ca-certificates",
        "openssl",
        "libssl1.0.2",
        "libssl1.1",
        "libbz2-1.0",
        "libdb5.3",
        "libffi6",
        "libncursesw5",
        "liblzma5",
        "libexpat1",
        "libreadline7",
        "libtinfo5",
        "libsqlite3-0",
        "mime-support",
        "netbase",
        "readline-common",
        "tzdata",

        #c++
        "libgcc1",
        "libgomp1",
        "libstdc++6",

        #java
        "zlib1g",

        #python
        "libpython2.7-minimal",
        "python2.7-minimal",
        "libpython2.7-stdlib",
        "dash",
        "libc-bin=2.28-10",

        #python3
        "libpython3.7-minimal",
        "python3.7-minimal",
        "libpython3.7-stdlib",
    ],
    
    sources = [
        # Since I could not find a buster/security source, I decided to go ahead with stretch/security. I am not sure if I might be creating conflicts with the packages here
        "@debian_stretch_security//file:Packages.json",
        "@debian_buster//file:Packages.json",
    ],
)
```  

### BUILD
```
load("@io_bazel_rules_docker//container:container.bzl", "container_push")
load("@io_bazel_rules_docker//container:container.bzl", "container_image")

container_image(
    name = "layer1",
    base = "@gcrbase//image:image.tar",
)


load("//package_manager:package_manager.bzl", "package_manager_repositories")
load("//package_manager:dpkg.bzl", "dpkg_list", "dpkg_src")

load("@package_bundle//file:packages.bzl", "packages")

container_image(
	name = "layer2",
	base = "layer1",
	debs = [
		packages["dash"],
        packages["libbz2-1.0"],
        packages["libc-bin"],
        packages["libexpat1"],
        packages["libdb5.3"],
        packages["libffi6"],
        packages["libncursesw5"],
        packages["libreadline7"],
        packages["libsqlite3-0"],
        packages["libssl1.1"],
        packages["libtinfo5"],
        packages["mime-support"],
        packages["readline-common"],
        packages["zlib1g"],
        packages["python2.7-minimal"],
        packages["libpython2.7-minimal"],
        packages["libpython2.7-stdlib"],

        packages["python3.7-minimal"],
        packages["libpython3.7-minimal"],
        packages["libpython3.7-stdlib"],

    ],
    entrypoint = [
    	"/usr/bin/python3.7",
    ],
    symlinks = {
    	"/usr/bin/python": "/usr/bin/python3.7",
    	"/usr/bin/python3": "/usr/bin/python3.7",
    },

)

container_push(
    name = "push",
    image = ":layer2",
    format = "Docker",
    registry = "index.docker.io",
    repository = "my/publicrepo",
)`

```
I am able to build an image out of it with but unable to launch Python 3.7. The error I get is the following:

`python: /lib/x86_64-linux-gnu/libc.so.6: version 'GLIBC_2.28' not found (required by python)`
`python: /lib/x86_64-linux-gnu/libc.so.6: version 'GLIBC_2.25' not found (required by python)`
`python: /lib/x86_64-linux-gnu/libc.so.6: version 'GLIBC_2.26' not found (required by python)`

Clearly, this is due to the base image having glibc-2.24 shared objects as opposed to the 2.28 required by Python 3.7.
Can anyone give me pointers so as to what workaround I can apply to get Python3.7 running on my distroless image? (It is very much possible that my approach to building a 3.7 image this way is flawed, in that case please explain how can I get it done) Or do I have to rebuild the base as well with the newer dependencies? Any help will be appreciated.

**Comment by kalexmills** [comment]
Related to #364

**Comment by HVP1621** [comment]
Finally I was able to get it working by RE-BUILDING the base image (`:base` tagged image) with Buster's dependencies. This solved the issue created by taking dependencies from different distros as now all packages were from a Buster snapshot. 
This obviously is not a stable way to create images which are production ready, but it served my purpose at the moment. 
I have built the image using the following dpkg_src rule and replaced the packages in the `WORKSPACE` file with the appropriate versions available with the Buster snapshot. 
```
dpkg_src(
    name = "debian_buster",
    arch = "amd64",
    distro = "buster",
    sha256 = "b0a3a1754ce592ec31463f2e893cd11e4f6914ea9d2a6f4bf53ab2e199d3ad2e",
    snapshot = "20190605T025513Z",
    url = "https://snapshot.debian.org/archive",
)
```
